src/Microsoft.ComponentDetection.Detectors/dockerfile/DockerfileComponentDetector.cs

```diff
@@ -0,0 +1,154 @@
+            {
+                var constructType = construct.GetType().ToString();
+                switch (constructType)
+                {
+                    case "Valleysoft.DockerfileModel.FromInstruction":
+                        baseImage = ParseFromInstruction(construct, escapeChar, stageNameMap);
+                        break;
+                    case "Valleysoft.DockerfileModel.CopyInstruction":
+                        baseImage = ParseCopyInstruction(construct, escapeChar, stageNameMap);
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            return baseImage;
+        }
+
+        private IDockerReference ParseFromInstruction(DockerfileConstruct construct, char escapeChar, Dictionary<string, string> stageNameMap)
+        {
+            var tokens = construct.Tokens.ToArray();
+            var resolvedFromStatement = construct.ResolveVariables(escapeChar).TrimEnd();
+            var fromInstruction = (Valleysoft.DockerfileModel.FromInstruction)construct;
+            string reference = fromInstruction.ImageName;
+            if (string.IsNullOrWhiteSpace(resolvedFromStatement) || string.IsNullOrEmpty(reference))
+            {
+                return null;
+            }
+
+            var stageName = fromInstruction.StageName;
+            stageNameMap.TryGetValue(reference, out var stageNameReference);
+
+            if (!string.IsNullOrEmpty(stageName))
+            {
+                if (!string.IsNullOrEmpty(stageNameReference))
+                {
+                    stageNameMap.Add(stageName, stageNameReference);
+                }
+                else
+                {
+                    stageNameMap.Add(stageName, reference);
+                }
+            }
+
+            if (!string.IsNullOrEmpty(stageNameReference))
+            {
+                if (HasUnresolvedVariables(stageNameReference))
+                {
+                    return null;
+                }
+
+                return DockerReferenceUtility.ParseFamiliarName(stageNameReference);
+            }
+
+            if (HasUnresolvedVariables(reference))
+            {
+                return null;
+            }
+
+            return DockerReferenceUtility.ParseFamiliarName(reference);
+        }
+
+        private IDockerReference ParseCopyInstruction(DockerfileConstruct construct, char escapeChar, Dictionary<string, string> stageNameMap)
+        {
+            string reference = string.Empty;
+            var resolvedFromStatement = construct.ResolveVariables(escapeChar).TrimEnd();
+            if (string.IsNullOrWhiteSpace(resolvedFromStatement))
+            {
+                return null;
+            }
+
+            return DockerReferenceUtility.ParseFamiliarName(reference);
+        }
+
+        private bool HasUnresolvedVariables(string reference)
+        {
+            return new Regex("[${}]").IsMatch(reference);
+        }
```

Comment:
I'm not sure you need this in the C# port. It was to work around limitations in node's regex.